maitake/src/wait/map.rs

```diff
@@ -204,6 +216,56 @@ pub struct Wait<'a, K: PartialEq, V> {
     waiter: Waiter<K, V>,
 }
 
+impl<'map, 'wait, K: PartialEq, V> Wait<'map, K, V> {
+    /// Returns a future that completes when the `Wait` item has been
+    /// added to the `WaitMap`, and is ready to receive data
+    ///
+    /// This is useful for ensuring that a receiver is ready before
+    /// sending a message that will elicit the expected response.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use std::sync::Arc;
+    /// use maitake::{scheduler::Scheduler, wait::map::{WaitMap, WakeOutcome}};
+    /// use futures_util::pin_mut;
+    ///
+    /// let scheduler = Scheduler::new();
+    /// let q = Arc::new(WaitMap::new());
+    ///
+    /// let q2 = q.clone();
+    /// scheduler.spawn(async move {
+    ///     let wait = q2.wait(0);
+    ///
+    ///     // At this point, we have created the future, but it has not yet
+    ///     // been added to the queue. We could immediately await 'wait',
+    ///     // but then we would be unable to progress further. We must
+    ///     // first pin the `wait` future, to ensure that it does not move
+    ///     // until it has been completed.
+    ///     pin_mut!(wait);
+    ///     wait.as_mut().enqueue().await.unwrap();
+    ///
+    ///     // We now know the waiter has been enqueued, at this point we could
+    ///     // send a message that will cause key == 0 to be returned, without
+    ///     // worrying about racing with the expected response, e.g:
+    ///     //
+    ///     // sender.send_with_id(0, SomeMessage).await?;
+    ///     //
+    ///     let val = wait.await.unwrap();
+    ///     assert_eq!(val, 10);
+    /// });
+    ///
+    /// assert!(matches!(q.wake(&0, 100), WakeOutcome::NoMatch(_)));
+    ///
+    /// let tick = scheduler.tick();
+    ///
+    /// assert!(matches!(q.wake(&0, 100), WakeOutcome::Woke));
+    /// ```
+    pub fn enqueue(self: Pin<&'wait mut Self>) -> EnqueueWait<'wait, 'map, K, V> {
```

Comment:
API-wise, it feels like kind of a bummer that this doesn't take the `Wait` by value and then return it, but it has to be pinned to be added to the queue, and can't move after it's enqueued...so we couldn't implement an API like that without boxing. so, this seems good :+1: